luaui/Widgets/map_start_position_suggestions.lua

```diff
@@ -0,0 +1,744 @@
+				role = teamStart.role and teamStart.role or nil,
+			}
+		end)
+	end)
+end
+
+local function selectModoptionConfigForPlayers(modoptionData, allyTeamCount, playersPerTeam)
+	Spring.Log(
+		widget:GetInfo().name,
+		LOG.INFO,
+		"Searching for start positions for " .. table.toString({
+			allyTeamCount = allyTeamCount,
+			playersPerTeam = playersPerTeam,
+		})
+	)
+
+	for _, teamConfig in ipairs(modoptionData.team) do
+		if teamConfig.playersPerTeam == playersPerTeam and teamConfig.teamCount == allyTeamCount then
+			Spring.Log(widget:GetInfo().name, LOG.INFO, "Found start positions")
+			return teamConfig.sides
+		end
+	end
+
+	return nil
+end
+
+local function loadStartPositions()
+	local modoptionDataRaw = Spring.GetModOptions().mapmetadata_startpos
+
+	if modoptionDataRaw == nil or string.len(modoptionDataRaw) == 0 then
+		Spring.Log(widget:GetInfo().name, LOG.WARNING, "No modoption start position data found")
+		return
+	end
+
+	local decoded = base64.Decode(modoptionDataRaw)
+	local decompressed = VFS.ZlibDecompress(decoded)
+	local parsed = Json.decode(decompressed)
+
+	local allyTeamCount, playersPerTeam = getTeamSizes()
+	local selectedConfig = selectModoptionConfigForPlayers(parsed, allyTeamCount, playersPerTeam)
+
+	if selectedConfig == nil then
+		Spring.Log(widget:GetInfo().name, LOG.WARNING, "Could not find matching start positions")
+		return
+	end
+
+	local positions = processModoptionTeamConfig(
+		parsed.positions,
+		selectedConfig
+	)
+
+	return positions
+end
+
+-- widget code
+-- ===========
+
+local font = nil
+local fontTutorial = nil
+
+---@type WidgetStartPositions
+local startPositions = {}
+
+---@type string
+local tooltipKey = nil
+
+---@type Timer
+local tooltipStartTime = 0
+
+local placedCommanders = {}
+
+local vsx, vsy = Spring.GetViewGeometry()
+
+local function glListCache(originalFunc)
+	local cache = {}
+
+	local function clearCache()
+		for key, listID in pairs(cache) do
+			gl.DeleteList(listID)
+		end
+		cache = {}
+	end
+
+	local function decoratedFunc(...)
+		local rawParams = { ... }
+		local params = {}
+		for index, value in ipairs(rawParams) do
+			if index > 1 then
+				table.insert(params, value)
+			end
+		end
+
+		local key = table.toString(params)
+
+		if cache[key] == nil then
+			local function fn()
+				originalFunc(unpack(params))
+			end
+			cache[key] = gl.CreateList(fn)
+		end
+
+		gl.CallList(cache[key])
+	end
+
+	local decoratedFunction = setmetatable({}, {
+		__call = decoratedFunc,
+		__index = {
+			invalidate = clearCache,
+			getCache = function()
+				return cache
+			end,
+			getListID = function(...)
+				local params = { ... }
+				local key = table.toString(params)
+				return cache[key]
+			end
+		}
+	})
+
+	return decoratedFunction
+end
+
+local function drawArrow(a, b, size, cStart, cEnd)
+	local dir = { b[1] - a[1], b[2] - a[2], b[3] - a[3] }
+	local length = math.sqrt(dir[1] ^ 2 + dir[2] ^ 2 + dir[3] ^ 2)
```

Comment:
microoptimisation opportunity
```suggestion
	local length = math.diag(dir[1], dir[2], dir[3])
```